File: connectors/clients/elixir/lib/verisim_client/octad.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

defmodule VeriSimClient.Octad do
  @moduledoc """
  Octad CRUD operations for VeriSimDB.

  Octads are the fundamental multi-modal entities in VeriSimDB. Each octad can
  carry data across all eight modalities (graph, vector, tensor, semantic,
  document, temporal, provenance, spatial). This module provides create, read,
  update, delete, and list operations.

  All functions take a `VeriSimClient.t()` as their first argument and return
  `{:ok, result}` or `{:error, reason}` tuples.

  ## Examples

      {:ok, client} = VeriSimClient.new("http://localhost:8080")

      {:ok, octad} = VeriSimClient.Octad.create(client, %{
        name: "My Entity",
        description: "A test octad",
        vector: %{embedding: [0.1, 0.2, 0.3], model: "ada-002"}
      })

      {:ok, fetched} = VeriSimClient.Octad.get(client, octad["id"])
  """

  alias VeriSimClient.Types

  @doc """
  Create a new octad entity.

  The server assigns a UUID and timestamps; the returned map contains the
  fully-populated record.

  ## Parameters

    * `client` — A `VeriSimClient.t()` connection.
    * `input`  — A `octad_input()` map. At minimum, `:name` should be set.
  """
  @spec create(VeriSimClient.t(), Types.octad_input()) ::
          {:ok, Types.octad()} | {:error, term()}
  def create(%VeriSimClient{} = client, input) when is_map(input) do
    VeriSimClient.do_post(client, "/api/v1/octads", input)
  end

  @doc """
  Retrieve a single octad by its unique identifier.

  ## Parameters

    * `client` — A `VeriSimClient.t()` connection.
    * `id`     — The octad UUID string.
  """
  @spec get(VeriSimClient.t(), String.t()) ::
          {:ok, Types.octad()} | {:error, term()}
  def get(%VeriSimClient{} = client, id) when is_binary(id) do
    VeriSimClient.do_get(client, "/api/v1/octads/#{id}")
  end

  @doc """
  Update an existing octad entity (partial update / merge semantics).

  Only the fields present in `input` are modified; omitted fields retain their
  current values.

  ## Parameters

    * `client` — A `VeriSimClient.t()` connection.
    * `id`     — The octad UUID string.
    * `input`  — A `octad_input()` map with the fields to update.
  """
  @spec update(VeriSimClient.t(), String.t(), Types.octad_input()) ::
          {:ok, Types.octad()} | {:error, term()}
  def update(%VeriSimClient{} = client, id, input)
      when is_binary(id) and is_map(input) do
    VeriSimClient.do_put(client, "/api/v1/octads/#{id}", input)
  end

  @doc """
  Delete a octad entity by its unique identifier.

  This is a hard delete — the entity and all associated modality data are
  removed. Provenance records are retained for auditability.

  ## Parameters

    * `client` — A `VeriSimClient.t()` connection.
    * `id`     — The octad UUID string.
  """
  @spec delete(VeriSimClient.t(), String.t()) :: :ok | {:error, term()}
  def delete(%VeriSimClient{} = client, id) when is_binary(id) do
    VeriSimClient.do_delete(client, "/api/v1/octads/#{id}")
  end

  @doc """
  List octad entities with pagination.

  ## Parameters

    * `client` — A `VeriSimClient.t()` connection.
    * `opts`   — Keyword list with optional `:limit` (default 20) and `:offset` (default 0).

  ## Examples

      {:ok, page} = VeriSimClient.Octad.list(client, limit: 50, offset: 100)
  """
  @spec list(VeriSimClient.t(), keyword()) ::
          {:ok, Types.paginated_response()} | {:error, term()}
  def list(%VeriSimClient{} = client, opts \\ []) do
    limit = Keyword.get(opts, :limit, 20)
    offset = Keyword.get(opts, :offset, 0)
    VeriSimClient.do_get(client, "/api/v1/octads?limit=#{limit}&offset=#{offset}")
  end
end

File: connectors/clients/zig/src/octad.zig
// SPDX-License-Identifier: MPL-2.0
// VeriSimDB Zig Client — Octad CRUD operations.

const std = @import("std");
const Client = @import("client.zig").Client;
const types = @import("types.zig");
const errors = @import("error.zig");

pub const ParsedOctad = std.json.Parsed(types.Octad);
pub const ParsedPage = std.json.Parsed(types.PaginatedResponse);

fn decodeOrError(
    comptime T: type,
    allocator: std.mem.Allocator,
    body: []const u8,
    status: u16,
    expected: []const u16,
) !std.json.Parsed(T) {
    for (expected) |s| if (s == status) {
        return std.json.parseFromSlice(T, allocator, body, .{ .ignore_unknown_fields = true }) catch
            return errors.ClientError.DecodeFailed;
    };
    const err = errors.parseServerError(allocator, body, status) catch
        return errors.ClientError.HttpError;
    err.deinit(allocator);
    return errors.ClientError.HttpError;
}

pub fn create(client: *Client, input: types.OctadInput) !ParsedOctad {
    const body = try std.json.stringifyAlloc(client.allocator, input, .{});
    defer client.allocator.free(body);
    const resp = try client.doPost("/api/v1/octads", body);
    defer resp.deinit(client.allocator);
    return decodeOrError(types.Octad, client.allocator, resp.body, resp.status, &.{201});
}

pub fn get(client: *Client, id: []const u8) !ParsedOctad {
    const path = try std.fmt.allocPrint(client.allocator, "/api/v1/octads/{s}", .{id});
    defer client.allocator.free(path);
    const resp = try client.doGet(path);
    defer resp.deinit(client.allocator);
    return decodeOrError(types.Octad, client.allocator, resp.body, resp.status, &.{200});
}

pub fn update(client: *Client, id: []const u8, input: types.OctadInput) !ParsedOctad {
    const body = try std.json.stringifyAlloc(client.allocator, input, .{});
    defer client.allocator.free(body);
    const path = try std.fmt.allocPrint(client.allocator, "/api/v1/octads/{s}", .{id});
    defer client.allocator.free(path);
    const resp = try client.doPut(path, body);
    defer resp.deinit(client.allocator);
    return decodeOrError(types.Octad, client.allocator, resp.body, resp.status, &.{200});
}

pub fn delete(client: *Client, id: []const u8) !void {
    const path = try std.fmt.allocPrint(client.allocator, "/api/v1/octads/{s}", .{id});
    defer client.allocator.free(path);
    const resp = try client.doDelete(path);
    defer resp.deinit(client.allocator);
    if (resp.status != 200 and resp.status != 204) {
        const err = try errors.parseServerError(client.allocator, resp.body, resp.status);
        err.deinit(client.allocator);
        return errors.ClientError.HttpError;
    }
}

pub fn list(client: *Client, page: u32, per_page: u32) !ParsedPage {
    const path = try std.fmt.allocPrint(
        client.allocator,
        "/api/v1/octads?page={d}&per_page={d}",
        .{ page, per_page },
    );
    defer client.allocator.free(path);
    const resp = try client.doGet(path);
    defer resp.deinit(client.allocator);
    return decodeOrError(types.PaginatedResponse, client.allocator, resp.body, resp.status, &.{200});
}
